File: solutions/csharp/food-chain/1/FoodChain.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

public static class FoodChain
{
    private static readonly string _verseBreak = "\n\n";
    
    private static readonly Dictionary<int, string> _animals = new() 
    {
        {1, "fly"},
        {2, "spider"},
        {3, "bird"},
        {4, "cat"},
        {5, "dog"},
        {6, "goat"},
        {7, "cow"},
        {8, "horse"}
    };
	
    private static readonly Dictionary<int, string> _secondLine = new() 
    {
        {1, string.Empty},
        {2, "It wriggled and jiggled and tickled inside her.\n"},
        {3, "How absurd to swallow a bird!\n"},
        {4, "Imagine that, to swallow a cat!\n"},
        {5, "What a hog, to swallow a dog!\n"},
        {6, "Just opened her throat and swallowed a goat!\n"},
        {7, "I don't know how she swallowed a cow!\n"},
        {8, "She's dead, of course!"}
    };
	
    public static string Recite(int verseNumber) =>
        FirstLine(verseNumber)
		+ SecondLine(verseNumber)
		+ MiddleLines(verseNumber)
		+ LastLine(verseNumber);

    public static string Recite(int startVerse, int endVerse) =>
		String.Join(_verseBreak,
		    Enumerable.Range(startVerse, endVerse)
			.Select(currentVerse => Recite(currentVerse))
		);

    private static string FirstLine(int verseNumber) =>
        $"I know an old lady who swallowed a {_animals[verseNumber]}.\n";

    private static string SecondLine(int verseNumber) => _secondLine[verseNumber];

    private static string MiddleLines(int verseNumber) =>
		(verseNumber == 8) ? string.Empty : 
		String.Concat(
		Enumerable.Range(2, verseNumber - 1)
			.Reverse()
			.Select(currentVerse => MiddleLine(currentVerse))
		);

    private static string MiddleLine(int verseNumber) =>
		$"She swallowed the {_animals[verseNumber]} to catch the {_animals[verseNumber - 1]}{SpiderBonus(verseNumber)}.\n";
	
	private static string SpiderBonus(int verseNumber) => 
		(verseNumber != 3) ? string.Empty : 
		" that wriggled and jiggled and tickled inside her";

    private static string LastLine(int verseNumber) =>
		(verseNumber == 8) ? string.Empty : 
        $"I don't know why she swallowed the fly. Perhaps she'll die.";
}

File: solutions/cpp/food-chain/1/food_chain.cpp
#include "food_chain.h"
#include <string>
#include <string_view>
#include <array>
namespace food_chain {
    namespace {
        constexpr std::string_view animals[] = {
            "", "fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"
        };
        constexpr std::string_view second[] = {
          "", "", "It wriggled and jiggled and tickled inside her.\n",
          "How absurd to swallow a bird!\n", "Imagine that, to swallow a cat!\n", 
          "What a hog, to swallow a dog!\n", "Just opened her throat and swallowed a goat!\n",
          "I don't know how she swallowed a cow!\n"
        };
        constexpr std::string_view ending = 
            "I don't know why she swallowed the fly. Perhaps she'll die.\n";
        std::string build_beginning(int verse) {
            return "I know an old lady who swallowed a " + std::string(animals[verse]) + ".\n";
        }
        std::string build_verse(int number) {
            std::string verse;
            verse += build_beginning(number);
            if (animals[number] == "horse") { // end the song immediately
                verse += "She's dead, of course!\n";
                return verse;
            }
            // add the second part
            verse += second[number];
            // now add "she swallowed the n to catch the n-1" part ONLY if you're not the fly (n = 1)
            if (number > 1) {
                for (int i = number; i > 1; --i) {
                    verse += "She swallowed the " + std::string(animals[i]) + " to catch the " + std::string(animals[i - 1]);
                    if (animals[i - 1] == "spider") { // for the bird
                        verse += " that wriggled and jiggled and tickled inside her";
                    }
                    verse += ".\n";
                }
            } 
            // add the ending for the fly
            verse += ending;
            return verse;
        } // end of build_verse()
    } // anonymous namespace
    std::string verse(int verse) {
        return build_verse(verse);
    }
    std::string sing() {
        return verses(1, 8);
    }
    std::string verses(int start, int end) {
        std::string result;
        for (int i = start; i <= end; ++i) { // start and end inclusive
            result += verse(i); // call the verse() method
            // add a blank line in between verses
            result += "\n";
        }
        return result;
    }
}  // namespace food_chain





